contracts/zaps/ZapCurveTokenToPrincipalToken.sol

```diff
@@ -0,0 +1,471 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.0;
+
+import "../libraries/Authorizable.sol";
+import "@openzeppelin/contracts/token/ERC20/extensions/draft-IERC20Permit.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/utils/Address.sol";
+import "../interfaces/IVault.sol";
+
+/// @title ZapCurveTokenToPrincipalToken
+/// @notice Allows the user to buy and sell principal tokens using a wider
+/// array of tokens
+/// @dev This contract introduces the concept of "root tokens" which are the set
+/// of constituent tokens for a given curve pool. Each principal token is
+/// constructed by a yield-generating position which in this case will be
+/// represented by a curve LP token. This is referred to as the "base token"
+/// and in the case where the user wishes to purchase or sell a principal token,
+/// it can only be done so by using this token.
+/// What this contract intends to do is enable the user purchase or sell a
+/// position using those "root tokens" which would garner significant UX
+/// improvements. The flow in the case of purchasing is as follows, the root
+/// tokens are added as liquidity into the correct curve pool, giving a curve "LP
+/// token" or "base token". Subsequently this is then used to purchase the
+/// principal token. Selling works similarly but in the reverse direction
+contract ZapCurveTokenToPrincipalToken is Authorizable {
+    using SafeERC20 for IERC20;
+    using Address for address;
+
+    // Store the accessibility state of the contract
+    bool public isFrozen = false;
+
+    // A constant to represent ether
+    address internal constant _ETH_CONSTANT =
+        address(0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE);
+
+    // Reference to the main balancer vault
+    IVault internal immutable _balancer;
+
+    /// @notice Marks the msg.sender as authorized and sets them as the owner
+    ///      in the authorization library
+    /// @param __balancer The balancer vault contract
+    constructor(IVault __balancer) Authorizable() {
+        _authorize(msg.sender);
+        _balancer = __balancer;
+    }
+
+    // Allows this contract to receive ether
+    receive() external payable {}
+
+    /// @notice This function sets approvals on all ERC20 tokens
+    /// @param tokens An array of token addresses which are to be approved
+    /// @param spenders An array of contract addresses, most likely curve and
+    /// balancer pool addresses
+    /// @param amounts An array of amounts for which at each index, the spender
+    /// from the same index in the spenders array is approved to use the token
+    /// at the equivalent index of the token array on behalf of this contract
+    function setApprovalsFor(
+        address[] memory tokens,
+        address[] memory spenders,
+        uint256[] memory amounts
+    ) external onlyAuthorized {
```

Comment:
Yes but the only issue with doing so would be where someone could set it to approval to 0 just to be chaotic. It could be possible to have an open approve if the amount they wish to set the contracts approval to is greater than the current

Reply:
It actually would be best to leave as authorized as it requires consideration of allowance behaviour across all tokens we would support.